File: pet-war/game/utils/circular_queue.js
class CircularQueue {
    constructor() {
        this.queue = [];
        this.size = 0;
    }

    shuffleSize(size) {
        for (let i = 0; i < size; i++) {
            const rand = Math.floor(Math.random() * size);
            [this.queue[i], this.queue[rand]] = [this.queue[rand], this.queue[i]];
        }
    }

    shuffleOn(start = 0, end) {
        if (end > this.queue.length) {
            end = this.queue.length;
        }
        // for (let i = start; i < end; i++) {
        //     const rand = Math.floor(Math.random() * (end - start + 1)) + (start - 1);
        //     [this.queue[i], this.queue[rand]] = [this.queue[rand], this.queue[i]];
        // }
        for (let i = start; i < end; i++) {
            // Generate a random index between i (inclusive) and end (exclusive)
            const rand = Math.floor(Math.random() * (end - i)) + i;
            // Swap elements at indices i and rand
            [this.queue[i], this.queue[rand]] = [this.queue[rand], this.queue[i]];
        }
    }

    shuffleAll() {
        for (let i = this.queue.length - 1; i > 0; i--) {
            const rand = Math.floor(Math.random() * (i + 1));
            [this.queue[i], this.queue[rand]] = [this.queue[rand], this.queue[i]];
        }
    }

    getAt(index) {
        return this.queue[index];
    }

    getRange(len) {
        if (len > this.queue.length) {
            len = this.queue.length;
        }
        return Array.from(this.queue).splice(0, len);
    }

    setElement(elements) {
        this.queue = elements;
        this.size = elements.length;
    }

    setAt(index, element) {
        this.queue[index] = element;
    }

    addElement(element) {
        this.queue.push(element);
        this.size += 1;
    }

    removeAt(index) {
        this.size -= 1;
        return this.queue.splice(index, 1);
    }

    copyAll(){
        return JSON.parse(JSON.stringify(this.queue));
    }

    moveForwardAll() {
        // console.log("Forward");
        let firstElement = this.queue[0];
        for (let i = 0; i < this.queue.length; i++) {
            this.queue[i] = this.queue[i + 1];
        }
        this.queue[this.queue.length - 1] = firstElement;
    }

    moveBackwardAll() {
        // console.log("Backward");
        let lastElement = this.queue[this.queue.length - 1];
        for (let i = this.queue.length - 1; i > 0; i--) {
            this.queue[i] = this.queue[i - 1];
        }
        this.queue[0] = lastElement;
    }

    printElement(len) {
        if (len > this.size) {
            len = this.size;
        }
        let temp = Array.from(this.queue).splice(0, len);
        // console.log(temp);
        // for (let i = 0; i < size; i++) {
        //     // console.log(this.queue[i]);
        // }
    }

    printAll() {
        // console.log(this.queue);
        // for (let i = 0; i < this.size; i++) {
        //     // console.log(this.queue[i]);
        // }
    }

    toArray() {
        return this.queue;
    }
}

module.exports = CircularQueue;

File: flutter_pet_war/lib/core/utils/circular_queue.dart
import 'dart:math';
// import 'dart:developer' as dev;

class CircularQueue<T> {
  List<T> data;
  int size;
  CircularQueue({this.data = const [], this.size = 0});

  void shuffleSize(int len) {
    for (var i = 0; i < len; i++) {
      var rand = Random().nextInt(i + 1);
      var temp = data[i];
      data[i] = data[rand];
      data[rand] = temp;
    }
  }

  void shuffleAll() {
    // for (var i = data.length - 1; i > 0; i--) {
    //   var rand = Random().nextInt(i + 1);
    //   var temp = data[i];
    //   data[i] = data[rand];
    //   data[rand] = temp;
    // }
    data.shuffle();
  }

  T getAt(int index) {
    return data[index];
  }

  List<T> getRange(int len) {
    if (len > data.length) {
      len = data.length;
    }
    return data.getRange(0, len).toList();
  }

  void setElement(List<T> elements) {
    data.removeWhere((element) => true);
    data.addAll(elements);
    size = data.length;
  }

  void addElement(T element) {
    data.add(element);
    size = data.length;
  }

  void addAll(Iterable<T> iterable) {
    data.addAll(iterable);
    size = data.length;
  }

  T removeAt(int index) {
    size = data.length;
    return data.removeAt(index);
  }

  void setAt(int index, T element) {
    data[index] = element;
    size = data.length;
  }

  void moveForwardAll() {
    // dev.inspect("Forward");
    var firstElement = data[0];
    for (var i = 0; i < data.length - 1; i++) {
      data[i] = data[i + 1];
    }
    data[data.length - 1] = firstElement;
  }

  void moveBackwardAll() {
    // dev.inspect("Backward");
    var lastElement = data[data.length - 1];
    for (var i = data.length - 1; i > 0; i--) {
      data[i] = data[i - 1];
    }
    data[0] = lastElement;
  }

  void printElement(int len) {
    if (len > size) {
      len = size;
    }
    // var temp = getRange(len);
    // dev.inspect(temp);
    for (var i = 0; i < size; i++) {
    }
  }

  void printAll() {
    // dev.inspect(data);
    // var text = "[";
    // for (var i = 0; i < size; i++) {
    //   text += "[${data[i]?.length ??}]";
    //   // print(data[i].toString());
    // }
    // text += "]";
    // print(text);
  }
}
